runtime/lua/vim/iter.lua

```diff
@@ -247,7 +249,7 @@ function ListIter:flatten(depth)
 
     -- exit early if we try to flatten a dict-like table
     if flattened == nil then
-      error('flatten() requires a list-like table')
+      error('flatten() requires an array-like table')
```

Comment:
Due to L135:
```
-- short-circuit: this is not a list like table
```

`flatten()` on lists with hole still might fail:

```lua
vim.iter({nil, {1, nil, 2}, 3}):flatten():totable()
-- ERROR: flatten() requires an array-like table
```

For the following, an expected output could be `{{1, nil, 2}, 3}` instead of an error according to the array-like semantics:
```
vim.iter({nil, {1, nil, 2}, 3}):flatten():totable()
-- ERROR: flatten() requires an array-like table
```

Reply:
Expected output of `vim.iter({nil, {1, nil, 2}, 3}):flatten():totable()` should be `{1, 2, 3}`.

`vim.iter({nil, {1, nil, 2}, 3}):totable()` (no flatten) returns `{{1, nil, 2}, 3}`. Calling flatten on that would result in `{1, 2, 3}`.

Reply:
Oh that's right. My mistake; I was actually thinking about `vim.iter(**{** {nil, {1, nil, 2}, 3} **}** ):flatten(1):totable()` (one more depth) == `{ {1, nil, 3}, 3 }`. So the current implementation after the update seems correct: no longer errors.